src/Microsoft.AspNetCore.Routing/CompositeEndpointDataSource.cs

```diff
@@ -60,11 +67,53 @@ private void Initialize()
         {
             lock (_lock)
             {
-                _changeToken = new CompositeChangeToken(_dataSources.Select(d => d.ChangeToken).ToArray());
+                _dataSourcesChangeToken = new CompositeChangeToken(_dataSources.Select(d => d.ChangeToken).ToArray());
                 _endpoints = _dataSources.SelectMany(d => d.Endpoints).ToArray();
 
-                _changeToken.RegisterChangeCallback((state) => Initialize(), null);
+                var oldChangeCallback = _dataSourcesChangeCallback;
+                _dataSourcesChangeCallback = _dataSourcesChangeToken.RegisterChangeCallback(
+                    (state) =>
+                    {
+                        // Refresh the endpoints from datasource so that callbacks can get the latest endpoints
+                        Initialize();
+
+                        // Prevent consumers from re-registering callback to inflight events as that can 
+                        // cause a stackoverflow
+                        // Example:
+                        // 1. B registers A
+                        // 2. A fires event causing B's callback to ge called
+                        // 3. B executes some code in its callback, but needs to re-register callback 
+                        //    in the same callback
+                        var oldTokenSource = _cts;
+                        var oldToken = _consumerChangeToken;
+
+                        CreateChangeToken();
+
+                        // Raise consumer callbacks. Any new callback registration would happen on the new token
+                        // created in earlier step.
+                        oldTokenSource.Cancel();
+
+                        oldTokenSource.Dispose();
+                    },
+                    state: null);
+
+                if (oldChangeCallback != null && ReferenceEquals(oldChangeCallback, _dataSourcesChangeCallback))
+                {
+                    oldChangeCallback.Dispose();
+                }
```

Comment:
I don't think this is needed. The Dispose here just tells you to stop listening. We don't need to stop listening because the old callback has already been fired. Did you copy this from somewhere else?